### PR TITLE
Add `update_transformers` to `BaseSynthesizer`

### LINKED ISSUE #1021: Add update_transformers to synthesizers
### Problem Description

_As a user, it would be helpful to have ways to manually set custom transformers to use on my data before modeling._

### Expected behavior

- Add  `update_transformers` method to `BaseSynthesizer`
- Parameters:
    - column_name_to_transformer (`dict`): A dictionary mapping the name of the column to the transformer instance. 
- Method should update the `HyperTransformer` based of the provided dict
- Validation:
    - Errors: (Raise if any 1 or more columns encounter the case. Do the checks first. We shouldn't partially update anything.)
        - Updating a transformer that is incompatible with the sdtype provided in the metadata
`Error: Column 'age' is a numerical column, which is incompatible with the 'LabelEncoder' preprocessing.`
        - Adding a transformer other than AnonymizedFaker or RegexGenerator for a key column (primary, alternate, sequence key)
`Error: Column 'user_id' is a key. It cannot be preprocessed using the 'FloatFormatter' transformer.`
        - The user is assigning a transformer object that has already been fit
`Error: Transformer for column 'age' has already been fit on data. `
    - Warnings: Raise all that arise
        - (CTGAN, CopulaGAN, TVAE, PAR only): Whenever the user tries to add a transformer for a column that is auto-assigned to None (boolean/categorical)
`Warning: Replacing the default transformer for column 'degree_type' might impact the quality of your synthetic data`
        - (GaussianCopula): Whenever the user is adding a OneHotEncoder to a categorical column
`Warning: Using the OneHotEncoder for column 'degree_type' may slow down the preprocessing and modeling time`

### PR DESCRIPTION
**Pull request by fealho**
Resolve #1021.

### CODE REVIEW

File: sdv/single_table/base.py [hunk 1]
```diff
@@ -234,3 +236,50 @@ def validate(self, data):
...
+                raise Error(
+                    f"Column '{column}' is a key. It cannot be preprocessed using "
+                    f"the '{transformer_name}' transformer."
+                )
+
+            # If columns were set, the transformer was fitted
+            if transformer.columns:
+                raise Error(f"Transformer for column '{column}' has already been fit on data.")
+
+    def _warn(self, column_name_to_transformer):
```
Comment (amontanez24) [inline sdv/single_table/base.py:261]: minor: maybe make this more descriptive. `_raise_transformer_update_warnings` or `_warn_for_transformer_updates`

File: sdv/single_table/copulas.py [hunk 2]
```diff
@@ -108,3 +108,18 @@ def _fit(self, processed_data):
...
+        """Raise warnings of a model.
+
+        Args:
+            column_name_to_transformer (dict):
+                Dict mapping column names to transformers to be used for that column.
+        """
+        for column in column_name_to_transformer:
+            sdtype = self.metadata._columns[column]['sdtype']
+            if sdtype == 'categorical':
+                warnings.warn(
```
Comment (amontanez24) [inline sdv/single_table/copulas.py:122]: I think there should be one more warning like this on the other models for when categorical  or boolean transformers are set

**Review by amontanez24 — CHANGES_REQUESTED** [review]

File: sdv/single_table/base.py [hunk 3]
```diff
@@ -234,3 +236,50 @@ def validate(self, data):
 
         if errors:
             raise InvalidDataError(errors)
+
+    def _validate_transformers(self, column_name_to_transformer):
+        keys = self._get_primary_and_alternate_keys() | self._get_set_of_sequence_keys()
+        for column, transformer in column_name_to_transformer.items():
+            transformer_name = type(transformer).__name__
+            sdtype = self.metadata._columns[column]['sdtype']
+            if sdtype != transformer.get_input_sdtype():
```
[the remark this answers is not in the document]
Comment (fealho) [reply sdv/single_table/base.py:245]: Good catch

**Review by amontanez24 — APPROVED** [review]
LGTM!

**Pull request merged**